Upgrade to svelte 5 and bump all other dependencies

ndom91 opened the pull request:
Hey man, was sooo happy to have found this when I got my niim labelmaker. As a web dev I want to help improve the UI, but first I figured this could use a nice little pass to ensure all eslint/prettier stuff is setup correctly and that all dependencies are up to date. Hope I haven't changed too much on you 😅 

This PR does the following:
- bump all dependencies
- ensure prettier/eslint are up to date and installed correctly
- format all files
- Add and enable sveltekit in "single page app" mode (client-side only)

MultiMote commented:
Wow, this PR is huge. I need some time to review it 😂

MultiMote commented:
Quick look:

1. Do we really need sveltekit? This is single page app without routes and server features. Using SvelteKit here is redundant; we only need a library for building a graphical interface.
One of my goals was to minimize dependencies, and now the project uses 411 dependencies instead of 265 🙂

2. Text block uses default font on first load. When you modify text (change size, for example) it becomes normal.
    <img width="719" height="202" alt="image" src="https://github.com/user-attachments/assets/eac42dd9-4e99-47ce-94ab-526e378a8943" />

3. Label saving to browser storage is now weird. Many labels saved, but one displayed.

MultiMote commented:
> Any ideas yourself or preferences?

Main goal is **not** make UI mobile-oriented 🙂
It should be comfortable both for PC and mobile users.
Currently my biggest pain is hi-res labels that not fit into screen. I think I should implement some pan-zoom functionality, but I am not ready to do that because it's complicated with fabric.js.

MultiMote commented:
What about eslint and prettier - I prefer to make them optional (project README has line about that), but having configs for them.

MultiMote commented on src/locale/dicts/cs.json:
Please revert locale files formatting, it generated by weblate

ndom91 commented on vite.config.ts:
```diff
@@ -9,10 +10,12 @@ const getDate = (): string => {
     __APP_VERSION__: JSON.stringify(process.env.npm_package_version),
-    __APP_COMMIT__: JSON.stringify(process.env.COMMIT_HASH),
+    __APP_COMMIT__: JSON.stringify(
+      execSync("git rev-parse HEAD").toString().trim().substring(0, 7),
```
What build environment doesnt have git installed? That's almost a prerequisite to getting this program 🤔 

But okay no problem

ndom91 replied:
Removed

ndom91 commented on src/fabric-object/qrcode.ts:
```diff
@@ -6,11 +6,46 @@ export type ErrorCorrectionLevel = "L" | "M" | "Q" | "H";
+  | 37
+  | 38
+  | 39
+  | 40;
```
Default formatting

ndom91 commented:
Okay, I've removed SvelteKit and addressed some of your feedback. To get to your last msg specifics:

--- 

> Quick look:
> 
> 1. Do we really need sveltekit? This is single page app without routes and server features. Using SvelteKit here is overkill; we only need a library for building a graphical interface.
>    One of my goals was to minimize dependencies, and now the project uses 411 dependencies instead of 265 🙂

Don't let JS hate on the internet get to you btw 😅, I see you're active in many non-web dev areas of software development. Of course keeping dependencies low generally is not a bad goal, but almost none of these would have gotten shipped to the client. We had basically only added `@sveltejs/kit` which gets compiled away for the mostt part during the build process anyway. Bigger additions to the client JS bundle size are things like `bootstrap` and `material-icons`.

> 2. Text block uses default font on first load. When you modify text (change size, for example) it becomes normal.
>    <img alt="image" width="719" height="202" src="https://private-user-images.githubusercontent.com/8068160/509181027-eac42dd9-4e99-47ce-94ab-526e378a8943.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI2OTQ5MDYsIm5iZiI6MTc2MjY5NDYwNiwicGF0aCI6Ii84MDY4MTYwLzUwOTE4MTAyNy1lYWM0MmRkOS00ZTk5LTQ3Y2UtOTRhYi01MjZlMzc4YTg5NDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMDlUMTMyMzI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWMwZWQ2YTIxZDA3OTA5OGJjNjM4MTRhMmQ2ZDIxZjRlZTc3ZjRlZjU1NWU3YzNlYzUzYjJiODA0NjFlM2ZiYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.C5fwPRoql4Fd3sjeZY4rNwECdqwZP4kGrLmJcOQmHGE">

Not sure exactly what you mean here. When I load it during dev (and clear all cache / localstorage / etc.), I get NotoSans by default and all other font things seem to work. Can you help me understand what the issue is here?

> 3. Label saving to browser storage is now weird. Many labels saved, but one displayed.

Can you help me reproduce this? Maybe was an issue only when we had sveltekit on. It seems to work as expected now 

<img width="1048" height="480" alt="CleanShot 2025-11-09 at 14 48 52" src="https://github.com/user-attachments/assets/77a19a07-3415-44d6-a322-f6443fa1b562" />


> 4. meta and some other tags in html template are missing.

SvelteKit would add those in the build process via stuff like this (see screenshot), you shouldn't put them in the root `app.html` itself, because we'd also want to potentially change them (based on the page you're on, or some page state, etc.). Anyway, I've removed SvelteKit now so we will have to handle that manually again.

<img width="444" height="231" alt="CleanShot 2025-11-09 at 14 47 50" src="https://github.com/user-attachments/assets/df2bab7e-a4c4-467e-8258-328786e20d23" />

MultiMote commented:
> You can't use "class" as prop because that's a keyword in js (i.e. for actual classes, so had to rename to className in the destructuring)

<img width="376" height="147" alt="image" src="https://github.com/user-attachments/assets/1e47c911-9008-47e4-b96b-956eef151f1e" />

Is this bad? No errors thrown.

ndom91 commented:
I've updated the class / `[key: string]` type issues. The issue isn't in defining a prop / type called "class" but using that in svelte then. Svelte will complain and give you cryptic error messages about cant find variable or something like that. That's why I've renamed it `className` in the destructuring from `$props`. 

Which btw that is new Svelte 5 sync, the `const { onItemDelete } = $props()` syntax. Is a lot easier to reason about rather than export let for each prop, one per line, imo.

Some of the tsconfig.json changes (like the `extends` was left-over from sveltekit. I've removed that now. The other change there, for example the `paths` ones, is so that you can use those as "shortcuts" to import stuff anywhere in the codebase. So for example, we have this entry in tsconfig.json now: `"$components/*": ["src/components/*"]`. That means that anywhere in your app now, you can import a component via that path. See for example in `dashboard.svelte`: 

<img width="526" height="103" alt="CleanShot 2025-11-12 at 19 59 01" src="https://github.com/user-attachments/assets/73f08954-6429-4fa6-9b78-3ca4284e40f2" />

This way you never have to worry about figuring out `../../../` (how many directories do I go up? And then what is the path of hte file I want again?)

To your last point, I also renamed `static/` back to `public/` and it copies your `.ico`, `webmanifest`, etc. into dist on build now too 👍

ndom91 commented:
Also just noticed that that PWABuilder site I linked earlier has a function to take your PWA and generate bundles to publish to the Microsoft Store, Apple App Store and Google Play Store 🤯. 

I tested the Play Store functionality with another side project of mine and it does successfully generate an APK that basically just wraps your PWA in a native app "cloak" and then provides all the config and what not you'd need to publish it on the Play Store too. 

Very cool, figured it might be of interest to you 👍 

https://www.pwabuilder.com/

<img width="763" height="479" alt="CleanShot 2025-11-12 at 20 16 25" src="https://github.com/user-attachments/assets/618fb139-94ba-4b8b-887f-a93ac68e8bcb" />

ndom91 commented:
I fixed all the remaining eslint issues (including the each keys) in that last commit. But give it a test please. I'm not sure where they're all used. 

Hopefully there's not much more remaining 😅. Appreciate you working with me on this!

MultiMote commented:
<img width="1425" height="171" alt="image" src="https://github.com/user-attachments/assets/b959480f-a9a2-4b54-9eb0-5a7d8045e55e" />

MultiMote commented on src/components/PrintPreview.svelte:
```diff
@@ -24,16 +24,20 @@
+  let printState: "idle" | "sending" | "printing" = $state("idle");
   let modal: Modal;
   let printProgress: number = 0; // todo: more progress data
   let density: number = $printerMeta?.densityDefault ?? 3;
```
Seems lile some state fields not converted to new state syntax (density/quantity/postProcessType and other). Is that how it was intended?

ndom91 commented on .DS_Store:
Nope. I'm new to mac os haha, more of a Linux desktop guy. Feel free to drop it

MultiMote commented:
Sorry for the slow response. I don't have much free time right now, and I'm spending it on other projects. Plus, it's not that easy to review such a large PR.

I can accept this PR into the test branch and slowly clean everything up myself, since I keep finding things I missed.

MultiMote commented:
Will do rest of things at https://github.com/MultiMote/niimblue/tree/svelte5-cleanup then shuash-merge to svelte5

MultiMote commented on src/components/LabelDesigner.svelte:
```diff
@@ -1,11 +1,13 @@
 <script lang="ts">
+  import { run } from "svelte/legacy";
```
Why `run` from `svelte/legacy` chosen instead of $effect?

ndom91 commented:
> You introduced $ path aliases, but migrated to it only partially.
> 
> And can you explain what the purpose of changes in tsconfig (except path aliases)?
> 
> <img alt="image" width="989" height="882" src="https://private-user-images.githubusercontent.com/8068160/518647179-0f9e9b6d-1650-4f70-83b9-b7862af4be2a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQwODA4MDgsIm5iZiI6MTc2NDA4MDUwOCwicGF0aCI6Ii84MDY4MTYwLzUxODY0NzE3OS0wZjllOWI2ZC0xNjUwLTRmNzAtODNiOS1iNzg2MmFmNGJlMmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjVUMTQyMTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTlhZDg2NzU4ZTFhZDQyNzFkYzA4ZTM0NjRjOTQyMTA5NjQ0ZjRjNWMxZjlhN2NjNzg3OTdlNDYzNDQzOTA5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.cbdloB0FdUDz6JPkABH_MDzNyn_EEpsuhYWrITnjWFY">

Yeah so I didnt change every import, but what else is missing? The path aliases are required (for non sveltekit projects) in `tsconfig.json`. The remaining changes were just the defaults when you create a new svelte 5 project that I'd copied over. Not strictly necessary if you dont want or like them. Just figured it'd be good to stay up-to-date with the recommendations from svelte

MultiMote commented on package.json:
```diff
@@ -6,46 +6,57 @@
-		"dev-open": "vite --force --open",
+		"dev": "vite",
+		"dev-open": "vite --open",
+		"build:android": "npm run build && npm run cap-run:android",
```
At that time, there was no capacitor project at root directory and you was added @capacitor/cli at root package.json. And script fails obviously.

MultiMote commented:
I think I finished it ([dev](https://github.com/MultiMote/niimblue/tree/dev) branch).
Deployed at https://dev.niim.blue, complete testing required,

MultiMote commented:
@ndom91 can you please take a look at current main brach code? I have no idea how to fix svelte5 reactivity problem. 

Problem is `selectedObject` state in `LabelDesigner`. When you edit properties in selected object, UI (TextParamsControls for example) is not updated because state is not deep reactive. Setting `selectedObject` to `undefined` then to previous value does not help. In svelte4 I solved this problem by reassigning (`selectedObject = selectedObject`).


https://github.com/user-attachments/assets/10486fc4-735d-4367-bdd4-4955b0a523fe
